File: backend/src/routes/disputes.ts
import { Router, Request, Response } from "express";
import { prisma } from "../config/db";
import { z } from "zod";

const router = Router();

const submitEvidenceSchema = z.object({
  submitted_by: z.string().min(1),
  content: z.string().optional().default(""),
  file_hash: z.string().optional().nullable(),
});

const createAppealSchema = z.object({
  requester_address: z.string().min(1),
});

const APPEAL_BUDGET_THRESHOLD = 10000000000; // 1000 USDC * 10^7

// GET /api/v1/disputes/:id
router.get("/:id", async (req: Request<{ id: string }>, res: Response) => {
  try {
    const { id } = req.params;
    const dispute = await prisma.disputes.findUnique({
      where: { id },
    });

    if (!dispute) {
      return res.status(404).json({ error: `dispute ${id} not found` });
    }

    res.json(dispute);
  } catch (error) {
    console.error("GET /disputes/:id error:", error);
    res.status(500).json({ error: "Internal server error" });
  }
});

// GET /api/v1/disputes/:id/evidence
router.get("/:id/evidence", async (req: Request<{ id: string }>, res: Response) => {
  try {
    const { id: disputeId } = req.params;

    const evidence = await prisma.evidence.findMany({
      where: { dispute_id: disputeId },
      orderBy: { created_at: "asc" },
    });

    res.json(evidence);
  } catch (error) {
    console.error("GET /disputes/:id/evidence error:", error);
    res.status(500).json({ error: "Internal server error" });
  }
});

// POST /api/v1/disputes/:id/evidence
router.post("/:id/evidence", async (req: Request<{ id: string }>, res: Response) => {
  try {
    const { id: disputeId } = req.params;
    const data = submitEvidenceSchema.parse(req.body);

    const evidence = await prisma.evidence.create({
      data: {
        dispute_id: disputeId,
        submitted_by: data.submitted_by,
        content: data.content,
        file_hash: data.file_hash,
      },
    });

    res.json(evidence);
  } catch (error) {
    if (error instanceof z.ZodError) {
      return res.status(400).json({ error: error.issues[0]?.message || "Validation failed" });
    }
    console.error("POST /disputes/:id/evidence error:", error);
    res.status(500).json({ error: "Internal server error" });
  }
});

// GET /api/v1/disputes/:id/verdict
router.get("/:id/verdict", async (req: Request<{ id: string }>, res: Response) => {
  try {
    const { id: disputeId } = req.params;

    const verdict = await prisma.verdicts.findFirst({
      where: { dispute_id: disputeId },
      orderBy: { created_at: "desc" },
    });

    if (!verdict) {
      return res.status(404).json({ error: "no verdict yet for this dispute" });
    }

    res.json(verdict);
  } catch (error) {
    console.error("GET /disputes/:id/verdict error:", error);
    res.status(500).json({ error: "Internal server error" });
  }
});

// POST /api/v1/disputes/:id/appeal
router.post("/:id/appeal", async (req: Request<{ id: string }>, res: Response) => {
  try {
    const { id: disputeId } = req.params;
    const data = createAppealSchema.parse(req.body);

    const dispute = await prisma.disputes.findUnique({
      where: { id: disputeId },
    });

    if (!dispute) {
      return res.status(404).json({ error: `dispute ${disputeId} not found` });
    }

    if (dispute.status !== "resolved") {
      return res.status(400).json({ error: "only resolved disputes can be appealed" });
    }

    const job = await prisma.jobs.findUnique({
      where: { id: dispute.job_id },
      select: { budget_usdc: true },
    });

    if (!job) {
      return res.status(404).json({ error: `job ${dispute.job_id} not found` });
    }

    if (Number(job.budget_usdc) < APPEAL_BUDGET_THRESHOLD) {
      return res.status(400).json({ error: `job budget (${job.budget_usdc} stroops) is below the appeal threshold (${APPEAL_BUDGET_THRESHOLD} stroops / 1000 USDC)` });
    }

    const existing = await prisma.appeals.findUnique({
      where: { dispute_id: disputeId },
    });

    if (existing) {
      return res.status(400).json({ error: "an appeal already exists for this dispute" });
    }

    const appeal = await prisma.appeals.create({
      data: {
        dispute_id: disputeId,
        status: "open",
      },
    });

    res.json(appeal);
  } catch (error) {
    if (error instanceof z.ZodError) {
      return res.status(400).json({ error: error.issues[0]?.message || "Validation failed" });
    }
    console.error("POST /disputes/:id/appeal error:", error);
    res.status(500).json({ error: "Internal server error" });
  }
});

export default router;

File: backend/src/routes/disputes.rs
use axum::{
    extract::{Path, State},
    routing::{get, post},
    Json, Router,
};
use uuid::Uuid;

use crate::{
    db::AppState,
    error::{AppError, Result},
    models::{Dispute, OpenDisputeRequest},
    routes::{appeals, evidence},
};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/:id", get(get_dispute))
        .route(
            "/:id/evidence",
            get(evidence::list_evidence).post(evidence::submit_evidence),
        )
        .route("/:id/verdict", get(crate::routes::verdicts::get_verdict))
        .route("/:id/appeal", post(appeals::create_appeal))
}

/// Open a dispute from within the job routes (/jobs/:id/dispute)
#[tracing::instrument(skip(state, req))]
pub async fn open_dispute_for_job(
    State(state): State<AppState>,
    Path(job_id): Path<Uuid>,
    Json(req): Json<OpenDisputeRequest>,
) -> Result<Json<Dispute>> {
    // Verify job is in a disputable state
    let status: Option<String> = sqlx::query_scalar("SELECT status FROM jobs WHERE id = $1")
        .bind(job_id)
        .fetch_optional(&state.pool)
        .await?;

    match status.as_deref() {
        Some("funded") | Some("in_progress") | Some("deliverable_submitted") => {}
        Some(s) => {
            return Err(AppError::BadRequest(format!(
                "cannot dispute job in status '{s}'"
            )))
        }
        None => return Err(AppError::NotFound(format!("job {job_id} not found"))),
    }

    // Update job status
    sqlx::query("UPDATE jobs SET status = 'disputed' WHERE id = $1")
        .bind(job_id)
        .execute(&state.pool)
        .await?;

    // Call escrow contract open_dispute via services::stellar
    let _ = state.stellar.open_dispute(&job_id.to_string()).await;

    let dispute = sqlx::query_as::<_, Dispute>(
        r#"INSERT INTO disputes (job_id, opened_by, status)
           VALUES ($1, $2, 'open')
           RETURNING id, job_id, opened_by, status, created_at"#,
    )
    .bind(job_id)
    .bind(req.opened_by)
    .fetch_one(&state.pool)
    .await?;

    Ok(Json(dispute))
}

#[tracing::instrument(skip(state))]
async fn get_dispute(
    State(state): State<AppState>,
    Path(dispute_id): Path<Uuid>,
) -> Result<Json<Dispute>> {
    let dispute = sqlx::query_as::<_, Dispute>(
        "SELECT id, job_id, opened_by, status, created_at FROM disputes WHERE id = $1",
    )
    .bind(dispute_id)
    .fetch_optional(&state.pool)
    .await?
    .ok_or_else(|| AppError::NotFound(format!("dispute {dispute_id} not found")))?;
    Ok(Json(dispute))
}

#[tracing::instrument(skip(state))]
pub async fn get_job_dispute(
    State(state): State<AppState>,
    Path(job_id): Path<Uuid>,
) -> Result<Json<Dispute>> {
    let dispute = sqlx::query_as::<_, Dispute>(
        r#"SELECT id, job_id, opened_by, status, created_at
           FROM disputes
           WHERE job_id = $1
           ORDER BY created_at DESC
           LIMIT 1"#,
    )
    .bind(job_id)
    .fetch_optional(&state.pool)
    .await?
    .ok_or_else(|| AppError::NotFound(format!("job {job_id} has no dispute")))?;

    Ok(Json(dispute))
}
